# CSC-6621: Final Project 

### Topic: Predicting Patient Readmission Rates using Random Forests and Gradient Boosting
**Team Members:** Sumedha Sanjeev, Ethan Traas, Ben Paulson<br/>
**Professor:** Dr. Berisha

In [34]:
import pandas as pd
import numpy as np
from tqdm import tqdm

## Part 1: Getting and Analyzing the Data
As mentioned in our official proposal for exploring this domain, this data comes from [data.cms.gov](https://data.cms.gov/provider-data/dataset/9n3s-kdb3#data-table), formally named "The Center of Medicare and Medicaid". The following description is provided by CMS as part of the Hospital Readmissions Reduction Program:
```text
In October 2012, CMS began reducing Medicare payments for subsection(d) hospitals with excess readmissions under the Hospital Readmissions Reduction Program (HRRP). Excess readmissions are measured by a ratio, calculated by dividing a hospital's predicted rate of readmissions for heart attack (AMI), heart failure (HF), pneumonia, chronic obstructive pulmonary disease (COPD), hip/knee replacement (THA/TKA), and coronary artery bypass graft surgery (CABG) by the expected rate of readmissions, based on an average hospital with similar patients.
```

The data is provided in a CSV file, which we will download and explore in the following `Part 1` section. It is a public dataset, and this is the data description:
* **Released:** December 10, 2020
* **Last Updated:** July 31, 2024
* **Rows:** 18774
* **Features (12):**
    * Facility Name
    * Facility ID
    * State
    * Measure Name
    * Number of Discharges
    * Footnote
    * Excess Readmission Ratio
    * Predicted Readmission Rate
    * Expected Readmission Rate
    * Number of Readmissions
    * Start Date
    * End Date



In [35]:
data = pd.read_csv('raw-data/FY_2024_Hospital_Readmissions_Reduction_Program_Hospital.csv')
data

Facility Name  Facility ID State  \
0       SOUTHEAST HEALTH MEDICAL CENTER        10001    AL   
1       SOUTHEAST HEALTH MEDICAL CENTER        10001    AL   
2       SOUTHEAST HEALTH MEDICAL CENTER        10001    AL   
3       SOUTHEAST HEALTH MEDICAL CENTER        10001    AL   
4       SOUTHEAST HEALTH MEDICAL CENTER        10001    AL   
...                                 ...          ...   ...   
18769  TRINITY REGIONAL HOSPITAL SACHSE       670319    TX   
18770  TRINITY REGIONAL HOSPITAL SACHSE       670319    TX   
18771  TRINITY REGIONAL HOSPITAL SACHSE       670319    TX   
18772  TRINITY REGIONAL HOSPITAL SACHSE       670319    TX   
18773  TRINITY REGIONAL HOSPITAL SACHSE       670319    TX   

                 Measure Name  Number of Discharges  Footnote  \
0      READM-30-HIP-KNEE-HRRP                   NaN       NaN   
1            READM-30-HF-HRRP                 616.0       NaN   
2           READM-30-AMI-HRRP                 274.0       NaN   
3            READM-30-PN-HRRP                 404.0       NaN   
4          READM-30-CABG-HRRP                 126.0       NaN   
...                       ...                   ...       ...   
18769  READM-30-HIP-KNEE-HRRP                   NaN       5.0   
18770        READM-30-HF-HRRP                   NaN       1.0   
18771      READM-30-COPD-HRRP                   NaN       5.0   
18772      READM-30-CABG-HRRP                   NaN       5.0   
18773       READM-30-AMI-HRRP                   NaN       5.0   

       Excess Readmission Ratio  Predicted Readmission Rate  \
0                        0.8916                      3.5325   
1                        1.1003                     23.1263   
2                        0.9332                     12.9044   
3                        0.9871                     17.0529   
4                        0.9517                      9.8131   
...                         ...                         ...   
18769                       NaN                         NaN   
18770                       NaN                         NaN   
18771                       NaN                         NaN   
18772                       NaN                         NaN   
18773                       NaN                         NaN   

       Expected Readmission Rate Number of Readmissions  Start Date  \
0                         3.9618      Too Few to Report  07/01/2019   
1                        21.0184                    149  07/01/2019   
2                        13.8283                     32  07/01/2019   
3                        17.2762                     68  07/01/2019   
4                        10.3112                     11  07/01/2019   
...                          ...                    ...         ...   
18769                        NaN                    NaN  07/01/2019   
18770                        NaN                    NaN  07/01/2019   
18771                        NaN                    NaN  07/01/2019   
18772                        NaN                    NaN  07/01/2019   
18773                        NaN                    NaN  07/01/2019   

         End Date  
0      06/30/2022  
1      06/30/2022  
2      06/30/2022  
3      06/30/2022  
4      06/30/2022  
...           ...  
18769  06/30/2022  
18770  06/30/2022  
18771  06/30/2022  
18772  06/30/2022  
18773  06/30/2022  

[18774 rows x 12 columns]

In [36]:
print(data.dtypes)

Facility Name                  object
Facility ID                     int64
State                          object
Measure Name                   object
Number of Discharges          float64
Footnote                      float64
Excess Readmission Ratio      float64
Predicted Readmission Rate    float64
Expected Readmission Rate     float64
Number of Readmissions         object
Start Date                     object
End Date                       object
dtype: object


In [37]:
print(data.describe())


         Facility ID  Number of Discharges     Footnote  \
count   18774.000000           8094.000000  6697.000000   
mean   261506.077661            244.251915     3.147230   
std    163743.437584            227.648660     2.084349   
min     10001.000000              0.000000     1.000000   
25%    110086.000000            105.000000     1.000000   
50%    250049.000000            176.000000     5.000000   
75%    390113.000000            308.000000     5.000000   
max    670319.000000           4994.000000     7.000000   

       Excess Readmission Ratio  Predicted Readmission Rate  \
count              12077.000000                12077.000000   
mean                   1.001776                   15.098235   
std                    0.077018                    5.480391   
min                    0.615900                    1.927900   
25%                    0.958000                   12.436800   
50%                    0.997800                   16.485500   
75%                    1.04

## Part 2: Preprocessing the Data
Because this data is supplied from a variety of sources across different hostpitals, it is not perfectly formatted yet in order to be used for machine learning. Therefore, we will need to preprocess the data before we can use it to train a model. The following steps will be taken to preprocess the data:
* General data cleaning
    * Remove leading "READM-30-" and trailing "-HRRP" from `Measure Name`
    * Make a new column `Length of Stay` from `Start Date` and `End Date`
* Handling missing/unhelpful values
    * Drop columns with too many missing values
    * Drop columns that are not useful (`Start Date`, `End Date` are all the same)
* Encode categorical variables
    * `Facility Name`, `State`, `Measure Name`
* Standardize numerical variables (0-1)
    * `Number of Discharges`, `Excess Readmission Ratio`, `Predicted Readmission Rate`, `Expected Readmission Rate`, `Number of Readmissions`
* Define the target variable
    * **Target:** `Predicted Readmission Rate`


#### Part 2.1: General Data Cleaning

In [38]:
# Cleaning the Measure Name column
data['Measure Name'] = data['Measure Name'].str.replace('READM-30-', '')
data['Measure Name'] = data['Measure Name'].str.replace('-HRRP', '')

# Determining the length of stay
data['Length of Stay'] = pd.to_datetime(data['End Date']) - pd.to_datetime(data['Start Date'])
data['Length of Stay'] = data['Length of Stay'].dt.days

#### Part 2.2: Remove Unhelpful Data
* `StartDate` and `EndDate` are the same, so we can drop one of them
* `Facility Name` is captured numerically by `Facility ID`

In [39]:
# Drop the `Start Date` and `End Date` and 'Facility Name' columns
data = data.drop(columns=['Start Date', 'End Date', 'Facility Name'])

#### Part 2.3: Encode Categorical Variables

In [40]:
# Encode categorical columns (Facility Name, State, Measure Name)
data = pd.get_dummies(data, columns=['State', 'Measure Name'])

#### Part 2.4: Standardize Numerical Variables

In [41]:
# Standardize numerical values; 0-1 (Number of Discharges, Excess Readmission Ratio, Predicted Readmission Rate, Expected Readmission Rate, Number of Readmissions)

# Convert any non-numeric values in above columns to NaN
data['Number of Discharges'] = pd.to_numeric(data['Number of Discharges'], errors='coerce')
data['Excess Readmission Ratio'] = pd.to_numeric(data['Excess Readmission Ratio'], errors='coerce')
data['Predicted Readmission Rate'] = pd.to_numeric(data['Predicted Readmission Rate'], errors='coerce')
data['Expected Readmission Rate'] = pd.to_numeric(data['Expected Readmission Rate'], errors='coerce')
data['Number of Readmissions'] = pd.to_numeric(data['Number of Readmissions'], errors='coerce')

# Standardize
data['Number of Discharges'] = (data['Number of Discharges'] - data['Number of Discharges'].min()) / (data['Number of Discharges'].max() - data['Number of Discharges'].min())
data['Excess Readmission Ratio'] = (data['Excess Readmission Ratio'] - data['Excess Readmission Ratio'].min()) / (data['Excess Readmission Ratio'].max() - data['Excess Readmission Ratio'].min())
data['Predicted Readmission Rate'] = (data['Predicted Readmission Rate'] - data['Predicted Readmission Rate'].min()) / (data['Predicted Readmission Rate'].max() - data['Predicted Readmission Rate'].min())
data['Expected Readmission Rate'] = (data['Expected Readmission Rate'] - data['Expected Readmission Rate'].min()) / (data['Expected Readmission Rate'].max() - data['Expected Readmission Rate'].min())
data['Number of Readmissions'] = (data['Number of Readmissions'] - data['Number of Readmissions'].min()) / (data['Number of Readmissions'].max() - data['Number of Readmissions'].min())

#### Part 2.5: Define the Target Variable

In [42]:
# Identify the target variable
data.rename(columns={'Predicted Readmission Rate': 'output'}, inplace=True)

# Drop rows where the target variable is NaN
data = data.dropna(subset=['output'])

#### Part 2.6: Determine What to do with NaN Values

`hit_list` is a list of the columns that have NaN values, as well as the percentage of that data which is NOT NaN
```text
('Number of Discharges', 65.33079407137534),
 ('Footnote', 0.0),
 ('Number of Readmissions', 65.33079407137534)]
 ```

 Depending on the model, we will either keep the NaN values or replace them with the average value of the column. For instance, one of the methods that will be tested (`XGBoost`) supports the interpretation of NaN.

In [43]:
# Identify which columns contain NaN values
hit_list = []
for column_name in list(data.columns):
    data_without_na = data.dropna(subset=[column_name])
    percentage = len(data_without_na) / len(data) * 100
    if percentage != 100:
        hit_list.append((column_name, percentage))

for hit in hit_list:
    if hit[1] < 50:
        data = data.drop(columns=[hit[0]])
    else:
        data[hit[0]] = data[hit[0]].fillna(data[hit[0]].mean())

#### Part 2.7: Get Data in State for SKLearn Models

In [44]:
# Format data for sklearn model training
data.to_csv('cleaned_data.csv', index=False)

In [45]:
len(data)

12077

## Part 3: Baseline Model - Linear Regression

The baseline model we will be using is Linear Regression. This model helps model the relationship between a dependent variable (output) and one or more independent variables (features) by fitting a linear equation to observed data. The goal is to predict the value of the dependent variable based on the values of the independent variables.

To train the linear regression model, the following steps will be taken:
- Split the data into training and testing sets
- Train th Linear Regression model:
    - Use the LinearRegression model from sklearn to fit the data.
    - The model learns the relationship between the features and the target variable by minimizing the sum of squared residuals.
- Hyperparameter Tuning: (**do we want this??**)
- Evaluate the Model:
    -Evaluate the model using Mean Squared Error (MSE) and R-squared (R²) metrics.
    - MSE is the average of the squares of the errors between the predicted and actual values. A smaller MSE indicates better model performance.
    - R² is the proportion of the variance in the dependent variable that is predictable from the independent variables. An R² value closer to 1 indicates a better fit.

In [46]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import mean_squared_error, r2_score

#### Part 3.1: Data Setup

In [47]:
data = pd.read_csv('cleaned_data.csv')
data.head()

Facility ID  Number of Discharges  Excess Readmission Ratio    output  \
0        10001              0.050174                  0.306231  0.061199   
1        10001              0.123348                  0.538043  0.808504   
2        10001              0.054866                  0.352438  0.418642   
3        10001              0.080897                  0.412307  0.576865   
4        10001              0.025230                  0.372987  0.300740   

   Expected Readmission Rate  Number of Readmissions  Length of Stay  \
0                   0.054545                0.046298            1095   
1                   0.777481                0.203240            1095   
2                   0.472732                0.030928            1095   
3                   0.618870                0.083947            1095   
4                   0.323661                0.000000            1095   

   State_AK  State_AL  State_AR  ...  State_WA  State_WI  State_WV  State_WY  \
0     False      True     False  ...     False     False     False     False   
1     False      True     False  ...     False     False     False     False   
2     False      True     False  ...     False     False     False     False   
3     False      True     False  ...     False     False     False     False   
4     False      True     False  ...     False     False     False     False   

   Measure Name_AMI  Measure Name_CABG  Measure Name_COPD  Measure Name_HF  \
0             False              False              False            False   
1             False              False              False             True   
2              True              False              False            False   
3             False              False              False            False   
4             False               True              False            False   

   Measure Name_HIP-KNEE  Measure Name_PN  
0                   True            False  
1                  False            False  
2                  False            False  
3                  False             True  
4                  False            False  

[5 rows x 64 columns]

In [48]:
# Split the data into training and testing sets
X = data.drop(columns=['output'])
y = data['output']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

#### 3.2 Train a Linear Regression Model

In [49]:
# Instantiate and train a Linear Regression Model
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)

LinearRegression()

In [50]:
linear_reg.coef_

array([-3.40724210e-09, -8.28392488e-02,  3.82199178e-01,  9.05645835e-01,
        9.85665598e-02,  4.72365203e-15, -5.22374168e-03, -7.76891126e-05,
        2.69272895e-03, -6.84809084e-04,  2.99683312e-03, -1.37187167e-03,
        2.33777635e-03,  1.04532425e-02, -7.22318553e-03,  1.10464772e-03,
       -7.45092740e-04,  2.27192386e-04, -1.95466536e-03, -5.54371367e-03,
        2.44166047e-03, -8.72403312e-04,  1.48946307e-04,  7.51527469e-04,
       -3.29229466e-04,  6.17800592e-03, -5.33767691e-03, -3.63048733e-03,
       -2.74914849e-05, -2.64799060e-03,  4.23418152e-07,  4.31886296e-03,
       -3.75840855e-03, -1.48435653e-03, -3.37054200e-03, -1.52423728e-03,
        4.73710699e-05,  1.52055843e-03, -2.47482958e-03,  5.36088264e-03,
        5.54081731e-03,  1.33305242e-03, -8.02456407e-05, -3.41836198e-03,
        1.52675860e-03,  1.89954348e-03,  1.75075994e-03, -4.11555362e-03,
        9.33011008e-04,  1.62744862e-03, -3.99643513e-03,  1.70425770e-03,
        2.90691320e-03, -

In [51]:
# Make model predictions
y_linreg_pred = linear_reg.predict(X_test)

In [52]:
# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Squared Error: {mse}')
print(f'R^2 Score: {r2}')

Mean Squared Error: 0.08020644821920594
R^2 Score: -0.9018546203690765


- Mean Squared Error (MSE): The value of `0.0802` suggests that the model's predictions have some error but isn't too high.
- R^2 Score: The value of `-0.902` indicates a poor model performance, suggesting that the linear regression model is not capturing the relationship between the features and the target variable effectively.

## Part 4: Random Forest Predictions

The improved and optimized model that we will be testing for this domain is the `Random Forest`. If you don't know what the Random Forest model is, you can read more about it [here](https://en.wikipedia.org/wiki/Random_forest). In summary, the Random Forest model is an ensemble learning method that operates by constructing a multitude of decision trees at training time and outputting the class that is the mode of the classes (classification) or mean prediction (regression) of the individual trees.

**The following steps will be taken to train and test the Random Forest model:**
* Split the data into training and testing sets
* Train the Random Forest model
    * Use the `RandomForestRegressor` model from `sklearn`
    * Hyperparameter `n_estimators` will be tuned. This hyperparameter is responsible for the number of trees in the forest. The larger the number of trees, the better the model will be but the longer it will take to run. Another downside of increasing the number of trees is that the model may overfit.
    * Hyperparameter `max_depth` will be tuned. This hyperparameter is responsible for the maximum depth of the tree. The deeper the tree, the more splits it has and it captures more information about the data. However, the deeper the tree, the more likely it is to overfit.
* Results will be evaluated
    * Output is `MSE` since this is a regression model. `MSE` is the mean squared error, which is the average of the squares of the errors. The larger the number, the larger the error. A good value to shoot for, given our output is a percentage, is 0.01 or less.

A random forest model was chosen because it is a robust model that can handle a large number of features and is less likely to overfit than a single decision tree. It is also a good model to use for regression problems because it handles non-linear relationships by the combination of the linear nature of decision trees.


In [53]:
# Import random forest from sklearn
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

#### Part 4.1: Data Setup

In [54]:
data = pd.read_csv('cleaned_data.csv')

In [55]:
# Split the data into training and testing sets
X = data.drop(columns=['output'])
y = data['output']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

#### Part 4.2: Train a Random Forest Model

In [56]:
# Instantiate and train a Random Forest Model
model = RandomForestRegressor(n_estimators=100, max_depth=10)
model.fit(X_train, y_train)

RandomForestRegressor(max_depth=10)

In [57]:
# Make predictions
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
print(mse)

1.4520815691539024e-05


#### Part 4.3: Hyperparameter Tune via Grid Search
Given we achieved an MSE of `1.1657932133341314e-05`, we can say that the Random Forest model is a good fit for this data. The next step is to try optimizing this model to achieve better results while visualizing how the mse changes given different parameters. We will use a simple two-level for-loop to iterate through different values of `n_estimators` and `max_depth` to find the best combination of hyperparameters.

The range of values which will be tested are as follows:
```python
n_estimators = [10, 25, 50, 100, 150, 200, 300, 400]
max_depth = [1, 2, 3, 4, 5, 10, 20, 30, 40, 50, 60]
```

In [58]:
n_estimators = [10, 25, 50, 100, 150, 200, 300, 400]
max_depth = [1, 2, 3, 4, 5, 10, 20, 30, 40, 50, 60]

results = []

for n in tqdm(n_estimators):
    for d in max_depth:
        model = RandomForestRegressor(n_estimators=n, max_depth=d)
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
        mse = mean_squared_error(y_test, y_pred)
        results.append((n, d, mse))

100%|██████████| 8/8 [10:40<00:00, 80.08s/it] 


In [59]:
# Determine the best hyperparameters
best = min(results, key=lambda x: x[2])
print(f"""
Best hyperparameters:
- n_estimators: {best[0]}
- max_depth: {best[1]}
- mse: {best[2]}
""")


Best hyperparameters:
- n_estimators: 200
- max_depth: 60
- mse: 1.2181948087228661e-05



In [60]:
# Visualize how the MSE changes with different hyperparameters (3D meshgrid, rotatable plot))
import plotly.graph_objects as go

# Create DataFrame
results_df = pd.DataFrame(results, columns=['n_estimators', 'max_depth', 'mse'])

# Prepare data for 3D plot
X = results_df['n_estimators']
Y = results_df['max_depth']
Z = results_df['mse']

# Create 3D plot
fig = go.Figure(data=[go.Mesh3d(x=X, y=Y, z=Z, color='lightblue', opacity=0.50)], layout=go.Layout(title='MSE vs. Hyperparameters'))

# Color the plot based on mse value
fig.add_trace(go.Scatter3d(x=X, y=Y, z=Z, mode='markers', marker=dict(size=12, color=Z, colorscale='Viridis', opacity=0.8)))

# Update layout
fig.update_layout(
    scene=dict(
        xaxis_title='Number of Estimators',
        yaxis_title='Max Depth',
        zaxis_title='Mean Squared Error'
    )
)

# Show plot
fig.show()

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

#### Part 4.4: Random Forest Conclusion

Based on the above results of Grid Search, we see that the best hyperparameters for the Random Forest model are `n_estimators=400` and `max_depth=30`. The MSE for this model is `8.782553467944984e-06`, which is a good value to shoot for given our output is a percentage. This means that the model is able to predict the readmission rate with an error of 0.0008782553467944984, which is a very small error on the scale of 0-1 that our output is on.

Evaluating why these parameters likely were the sweet-spot for accuracy while balancing the number of trees and branches within those trees, I believe that the model was able to capture the complexity of the data while not overfitting. The model was able to generalize well to the test data and predict the readmission rate with a very small error.


## Part 5: Gradient Boosting Predictions

Beyond `Random Forest`, another great application of machine learning to regression problems is `Gradient Boosting`. If you don't know what Gradient Boosting is, you can read more about it [here](https://en.wikipedia.org/wiki/Gradient_boosting). In summary, Gradient Boosting is a machine learning technique for regression and classification problems that produces a prediction model in the form of an ensemble of weak prediction models, typically decision trees. It builds the model in a stage-wise fashion like other boosting methods do, and it generalizes them by allowing optimization of an arbitrary differentiable loss function.

To summarize the difference between `Gradient Boosting` and `Random Forests`, the main difference is that `Gradient Boosting` builds trees sequentially, where each tree tries to correct the mistakes of the previous one. On the other hand, `Random Forests` builds trees independently, where each tree is built from a sample of the data.

This section will be an evaluation similar to that seen in `Part 2`, allowing comparison of the two approaches to this domain.

In [ ]:
# TODO

# NOTE: If these models are too good, could make a bigger split
# NOTE: When testing, you should have a validation set
# NOTE: Could try the model without doing feature engineering
# NOTE: After gradient boosting, could do a simple neural network (non-deep) to see if we even need the complex deep learning models